### PR TITLE
Fix missing statistics

### PR DESCRIPTION
**Pull request by LTMeyer**
This PR addresses #41.

# Statistics Updates

Statistics have been updated in the repo, on the public endpoint, on Hugging Face. 
The statistics computation now uses scientific notation with 4 decimal precision. This avoids insignificant changes in statistics when re-running the script.

# Test Updates

This PR reorganizes dataset tests. 
It tests the download of two datasets namely `active_matter` and `turbulent_radiative_layer_2D` to verify there is no issue, as it was the case with missing statistics. ~This makes the test significantly slower.~ (Tests may be slower locally based on network bandwidth.) The download is performed only once but used in several tests thanks to pytest fixture.

### CODE REVIEW

File: the_well/data/datasets.py [hunk 1]
```diff
@@ -276,26 +276,12 @@ def __init__(
 
         # Initialize normalization classes if True
         if use_normalization and normalization_type:
-            try:
-                with self.fs.open(self.normalization_path, mode="r") as f:
-                    stats = yaml.safe_load(f)
+            with self.fs.open(self.normalization_path, mode="r") as f:
```
Comment (LTMeyer) [inline the_well/data/datasets.py:279]: Remove the large try-except block to avoid silent failure.

File: scripts/compute_statistics.py [hunk 2]
```diff
@@ -140,11 +149,19 @@ def compute_statistics(train_path: str, stats_path: str):
...
     data_dir = args.the_well_dir
-
-    for dataset in WELL_DATASETS:
-        compute_statistics(
-            train_path=os.path.join(data_dir, dataset, "data/train"),
-            stats_path=os.path.join(data_dir, dataset, "stats.yaml"),
-        )
+    n_processes = args.n
+
+    with mp.Pool(n_processes) as pool:
```
Comment (LTMeyer) [inline scripts/compute_statistics.py:157]: Make statistics computation parallel.

**Comment by payelmuk150** [comment]
> In addition I think this delta normalization breaks when using a `dt_stride` than 1.
@francois-rozet , yeah the DeltaWellDataset normalization statistics were calculated for next step deltas. @mikemccabe210, should we add a flag for this? because there will be different normalization stats for different dt_strides. wdyt?

**Comment by francois-rozet** [comment]
> yes for the DeltaWellDataset, the input and target normalizations are indeed handled separately. Look at the _process_field_data function in DeltaWellDataset.

Oh right, i didnt't notice that the trajectory was split then concatenated again.

Note that

https://github.com/PolymathicAI/the_well/blob/8bebdb369c351e7ad5db6b2632020f49e7e79674/the_well/data/datasets.py#L873-L875

is equivalent to

```python
x = field_data[: self.n_steps_input]
y = field_data[self.n_steps_input - 1 :]
```

**Comment by francois-rozet** [comment]
> were calculated for next step deltas (which is what used for all the Well benchmarking)

Wait. Does this mean that the VRMSE in the tables in The Well are computed on the deltas, not on the fields?

**Comment by payelmuk150** [comment]
> > were calculated for next step deltas (which is what used for all the Well benchmarking)
> 
> Wait. Does this mean that the VRMSE in the tables in The Well are computed on the deltas, not on the fields?

@francois-rozet , no all the values in the paper are on the full fields. Delta option has been added recently much after the paper, so answer to your question is no. 

In any case, in the current repo, for both delta as well as non-delta case, the validation loss is calculated on the reconstructed prediction of the full field.

**Comment by mikemccabe210** [comment]
@francois-rozet no, I think Payel meant MPPX though we're still computing both over there - the training objective is just the delta version.

and @payelmuk150 that's a good point. Maybe we should rename it specifically to `lagk_...` and throw an error if the particular lag isn't in the stats file. We would then also need to provide some guidance on how the user would generate their own lags and update the compute stats to allow that.

File: scripts/compute_statistics.py (hunk 2 above)
Comment (mikemccabe210) [reply scripts/compute_statistics.py:157]: My assumption has been the reason this wasn't originally parallelized is because some of the datasets are quite large and would generally lead to OOM on typical machines, though maybe @francois-rozet could confirm if that's actually the case. I think @payelmuk150 might have also tried and hit memory issues on whatever machine she was running it on and our machines have higher than average memory. 

(though we could probably strictly chunk the computation to reduce this issue)

**Comment by francois-rozet** [comment]
> throw an error if the particular lag isn't in the stats file.

That's why I was saying it would be better to have a separate stats file for the deltas. Then you can have a different file for each lag/stride.

**Comment by payelmuk150** [comment]
> @francois-rozet no, I think Payel meant the other project though we're still computing both over there - the training objective is just the delta version.
> 
> and @payelmuk150 that's a good point. Maybe we should rename it specifically to `lagk_...` and throw an error if the particular lag isn't in the stats file. We would then also need to provide some guidance on how the user would generate their own lags and update the compute stats to allow that.

This will need a few changes. 

1) Make a stats_lag1 file and transfer all the current delta stats to there. 
2) Add guidance on how to generate additional lag stats files.
3) Massage the datasets.py and normalization.py script to work with changed file names

**Comment by mikemccabe210** [comment]
@francois-rozet Honestly, I'm not opposed to either path, but it's minor enough to me that I'm happy to let whoever gets around to implementing it choose.

Different files:
- Positives - More obviously modular. Avoids recomputing more stats than needed. Avoids needing specific field naming schemas (though yaml headers will always need to match the `Normalization` object's schema)
- Negatives - either need fixed naming schema (which makes extension harder) or users now need to explicitly pass stats files locations.

Same file:
- Positives - Less files hanging around/needing to be copied around. Less risk of accidentally overwriting one file with another that has the same name (because someone forgot to change the name while running the stats script) but different info. 
- Negatives - fields would now need a strict naming schema. You'd (for now) need to recompute all fields every time any time you want to add a new normalization type (unless we added an append mode to the stats script). Not being sure which fields you have in a stats file could get very annoying.

File: scripts/compute_statistics.py (hunk 2 above)
Comment (LTMeyer) [reply scripts/compute_statistics.py:157]: By default the number of processes is set to one. And we can also edit this script to target only one dataset. I think it is quite minor for the frequency with which this script is run.

**Comment by LTMeyer** [comment]
If we have one statistics file per possible stride, I'm afraid it will become quite intractable easily. I would rather allow instantiating normalization from values directly. These values could be stored in files, but we would not provide values for all length of strides.

I agree the points raised by @francois-rozet are valid. However, I believe they require some larger re-working that falls beyond the scope of this PR, which was fixing the current silent error reported in #44.
In the meantime, we can raise an error or a warning if the stride is incompatible with the computed statistics. What do you think about this?

**Comment by LTMeyer** [comment]
>     1. Make a stats_lag1 file and transfer all the current delta stats to there.
> 
>     2. Add guidance on how to generate additional lag stats files.
> 
>     3. Massage the datasets.py and normalization.py script to work with changed file names

I like these points and it will be a good solution. Especially the re-working of the normalization. It could be something like
```python

class Normalization:
    def __init__(self):
      # Init values required to normalize

    @classmethod
    def from_file(cls, file_name):
       # Create an instance from a statistics file

    def ___call___(self, x):
       # Perform the normalization on x input
```

I believe this would be some decent amount of changes though and may be better suited in a dedicated PR.

**Review by payelmuk150 — APPROVED** [review]
